### ISSUE TITLE
-Wundefined-var-template could use fixits

### ISSUE THREAD

**Issue by ben-craig** [llvm/llvm-project#62121]
The syntax for explicitly instantiating a template declaration is not easy to guess.  In addition, this is a warning that users are likely to get when migrating from gcc or msvc, so the "fixer" of this warning is very likely to be different than the originator of the code.  So we should help them out as much as we can.

This may want to even include _multiple_ fixits, to cover the varying use cases.

So if we have code like the following:
```
template <typename T>
struct has_static {
    static std::string my_str;
};
```
Then we may want one fixit for the "generic" case `template <typename T> std::string has_static<T>::my_str;`.

And another for the explicitly instantiated in a .cpp file case `template <> std::string has_static<int>::my_str;`.

Then there's yet another syntax, which I can't recall the pros / cons for: `extern template std::string has_static<int>::my_str;`.

**Comment by diohabara** [comment]
Hi, please assign this one.

**Comment by inclyc** [comment]
> Hi, please assign this one.

Thanks for contributing :) Please CC me if you submit patches!

**Comment by diohabara** [comment]
@ben-craig

Let me ask a question because I do not understand the issue well.

Is it possible to give three separate warnings with a same input? In this case, given:

```
template <typename T>
struct has_static {
    static std::string my_str;
};
```

Do you mean that I should give three separate warning with explanation at the same time like below?

```
template <typename T>
struct has_static {
    static std::string my_str;
};

This code can be rewritten as

1. for generic case, you can explicitly instantiate a template by specifying the type in the definition:
template <typename T>
std::string has_static<T>::my_str;

2. you can explicitly instantiated using template <> sytanx:
template <> std::string has_static<int>::my_str;

3. you can specify that explicit instantiation of the tempalte will be provided elsewhere
extern template std::string has_static<int>::my_str;
```

**Comment by ben-craig** [comment]
I would be fine with something like that.  Each of those three could also plausibly go in a separate note.  For the typed instantiations, you can fill in the type from the point of usage.
